### ISSUE TITLE
Query block missing first post when multiple query blocks are used on a single template

### ISSUE THREAD

**Issue by graylaurenm** [WordPress/gutenberg#53593]
### Description

When customizing a category template with a custom query block (above the default, "inherit query from template" query) the default query loop is missing the first post. Note that this happens whether or not the first query block is "custom" (or inherits the query).

When navigating to the second page, this continues.

For instance, if the query originally showed:

* Page 1 = Post 1, Post 2, Post 3
* Page 2 = Post 4, Post 5, Post 6

Then after adding another query it will instead show:

* Page 1 = Post 2, Post 3
* Page 2 = Post 5, Post 6

### Step-by-step reproduction instructions

I've included the full template code, where my additions start with the "custom query" paragraph block and end with the "default query" paragraph block.

```html
<!-- wp:template-part {"slug":"header","theme":"twentytwentythree","tagName":"header"} /-->

<!-- wp:group {"tagName":"main","style":{"spacing":{"margin":{"top":"var:preset|spacing|70","bottom":"var:preset|spacing|70"}}},"layout":{"type":"constrained"}} -->
<main class="wp-block-group" style="margin-top:var(--wp--preset--spacing--70);margin-bottom:var(--wp--preset--spacing--70)"><!-- wp:query-title {"type":"archive","align":"wide","style":{"spacing":{"margin":{"bottom":"var:preset|spacing|50"}}}} /-->

<!-- wp:paragraph -->
<p>custom query</p>
<!-- /wp:paragraph -->

<!-- wp:query {"queryId":12,"query":{"perPage":3,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
<div class="wp-block-query"><!-- wp:post-template -->
<!-- wp:columns {"verticalAlignment":"center"} -->
<div class="wp-block-columns are-vertically-aligned-center"><!-- wp:column {"verticalAlignment":"center","width":"25%"} -->
<div class="wp-block-column is-vertically-aligned-center" style="flex-basis:25%"><!-- wp:post-featured-image {"isLink":true} /--></div>
<!-- /wp:column -->

<!-- wp:column {"verticalAlignment":"center","width":"75%"} -->
<div class="wp-block-column is-vertically-aligned-center" style="flex-basis:75%"><!-- wp:post-title {"isLink":true} /--></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
<!-- /wp:post-template --></div>
<!-- /wp:query -->

<!-- wp:paragraph -->
<p>default query</p>
<!-- /wp:paragraph -->

<!-- wp:query {"queryId":2,"query":{"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":true,"perPage":3},"align":"wide","layout":{"type":"default"}} -->
<div class="wp-block-query alignwide"><!-- wp:post-template {"align":"wide","layout":{"type":"grid","columnCount":3}} -->
<!-- wp:post-featured-image {"isLink":true,"width":"100%","height":"clamp(15vw, 30vh, 400px)","align":"wide"} /-->

<!-- wp:post-title {"isLink":true} /-->

<!-- wp:post-excerpt /-->

<!-- wp:post-date {"isLink":true} /-->

<!-- wp:spacer {"height":"var(\u002d\u002dwp\u002d\u002dpreset\u002d\u002dspacing\u002d\u002d50)"} -->
<div style="height:var(--wp--preset--spacing--50)" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->
<!-- /wp:post-template -->

<!-- wp:query-pagination {"paginationArrow":"arrow","layout":{"type":"flex","justifyContent":"space-between"}} -->
<!-- wp:query-pagination-previous {"label":"Newer Posts"} /-->

<!-- wp:query-pagination-next {"label":"Older Posts"} /-->
<!-- /wp:query-pagination --></div>
<!-- /wp:query --></main>
<!-- /wp:group -->

<!-- wp:template-part {"slug":"footer","theme":"twentytwentythree","tagName":"footer"} /-->
```

### Screenshots, screen recording, code snippet

For example, on a fresh install with no plugins and using TT3, I set Settings > Reading > Blog pages show at most to "3" and can see the category template looks like this: 
![ss-2023-08-11-17 17 11](https://github.com/WordPress/gutenberg/assets/1750577/fb0b4857-d80c-4674-847c-4e3bea11810a)

If I add another query to the template, however, the first post is missing: 
![ss-2023-08-11-17 18 26](https://github.com/WordPress/gutenberg/assets/1750577/ecb6c145-9865-4085-baf9-f8b9a968ef07)

### Environment info

WordPress 6.3 running Twenty Twenty-Three theme.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by ndiego** [comment]
I have been able to confirm this issue. I copied the template code provided above using the TT3 theme in WordPress 6.3.

**Comment by ZafarKamal123** [comment]
Rewinding the global query just before it's processed seems to be fixing this issue. I have coded a [one file plugin gist](https://gist.github.com/ZafarKamal123/4087f9e7291d95326dd59eb20ac5702b) that seems to be fixing this issue.

**Comment by gziolo** [comment]
@felixarntz, it would be great to cross-check whether https://github.com/WordPress/gutenberg/pull/49904 fixed this issue, too. Based on the comment from @ZafarKamal123, it's very possible.

**Comment by felixarntz** [comment]
@gziolo Yes, it should be fixed by #49904 indeed. In fact, it is caused by the same problem that also causes https://core.trac.wordpress.org/ticket/58027, because of the hack `if ( ! in_the_loop() && have_posts() ) { the_post(); }` in the `core/post-featured-image` and `core/post-content` blocks:
* The custom query block will cycle through some posts not related to the main query, displaying the featured image for them.
* The problematic logic in the featured image block then triggers the main query loop, in a place where we're not supposed to be in the main query loop.
* When later the actual main query loop is encountered, it was actually already started by the one `the_post()` call, that's why the posts start at the 2nd one.

In other words, this issue is another great example for why those workarounds were problematic. 👍

**Comment by gziolo** [comment]
I can confirm it works as expected in the recent version of `trunk`.

Category page with the template included in the description:

<img width="1076" alt="Screenshot 2023-09-01 at 12 41 57" src="https://github.com/WordPress/gutenberg/assets/699132/cdde829d-176a-4827-91d5-9eb7b18144bc">

Category (2nd) page:

<img width="1249" alt="Screenshot 2023-09-01 at 12 42 06" src="https://github.com/WordPress/gutenberg/assets/699132/78098ca6-e075-40dc-b13d-b10c985eb05d">
